### PR TITLE
Add foreign key constraint support to SQLite3 adapter, fixed collation parsing

### PR DESCRIPTION
**Pull request by DavidMikeSimon**
The biggest headache with this turned out to be getting the constraint's name. SQLite3 provides `PRAGMA foreign_key_list` for reflecting on the foreign keys defined in the database, but the information it returns does not include the names of the constraints! At least as of the current SQLite version, the only way to retrieve that name is to parse the recorded `CREATE TABLE` statement from `sqlite_master`.

This is similar to an older problem with the SQLite3 adapter, where it was necessary to parse the `CREATE TABLE` statement to retrieve information about text column collation. The current implementation uses a simple regex to retrieve the collation name. I wanted to follow the same pattern, but the existing code was very weak to unusual column names (e.g. those with escaped quotes) and to unexpected syntax (e.g. if the column names were **not** quoted). Therefore I replaced that code with the barest minimum of a tokenizer/parser system.

Then, I added to that the ability to parse the foreign key `REFERENCES` statements. SQLite3 annoyingly allows those be defined in two different places with slightly different syntax...

I also added some tests to exhibit the bug in the existing collation code, and some tests to make sure the new parser code worked correctly in various border cases.

### CODE REVIEW

**Comment by DavidMikeSimon** [comment]
Oh, one additional note: in my code, the SQLite temporary database is no longer used for holding data while tables are recreated with schema changes, as any foreign key references would fail during that interim. This may have some performance impact, but AIUI it should be minimal, since SQLite is mostly used in Rails for development, and schema changes are relatively rare events anyways.

**Comment by DavidMikeSimon** [comment]
Looks like the foreign key violation error messages coming from SQLite on Travis are slightly different from what I'm getting on rails-dev-box. I've adjusted `translate_exception` a bit to handle both varieties, tests should pass now...

**Comment by DavidMikeSimon** [comment]
@sgrif Bump.

**Comment by matthewd** [comment]
> a tokenizer/parser system

:-1: do not want.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 1]
```diff
@@ -171,6 +173,10 @@ def supports_explain?
         true
       end
 
+      def supports_foreign_keys?
+        true
+      end
+
```
Comment (smathy) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Put a version test here so you're not returning `true` erroneously for older versions that don't support it.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 2]
```diff
@@ -361,15 +387,7 @@ def columns(table_name) #:nodoc:
...
-          sql = <<-SQL
-            SELECT sql
-            FROM sqlite_master
-            WHERE name=#{quote(row['name'])} AND type='index'
-            UNION ALL
-            SELECT sql
-            FROM sqlite_temp_master
-            WHERE name=#{quote(row['name'])} AND type='index'
-          SQL
+          sql = "SELECT sql FROM sqlite_master WHERE name=#{quote(row['name'])} AND type='index'"
```
Comment (smathy) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Can you explain why this change is part of this PR?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb [hunk 3]
```diff
@@ -333,6 +333,22 @@ def foreign_key(table_name, options = {}) # :nodoc:
...
+          foreign_keys.delete_if do |table_name, options|
+            options[:column].to_s == options_or_to_table[:column].to_s
+          end
+        elsif options_or_to_table.is_a?(String) or options_or_to_table.is_a?(Symbol)
+          foreign_keys.delete options_or_to_table.to_s
+        else
+          raise ArgumentError, "remove_foreign_key needs either a table name or a foreign key options hash with :name or :column"
+        end
+      end
+
```
Comment (timbreitkreutz) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb]: I believe the rails style guide suggests using `&&` and `||` and braces on `.delete()` ?

Also I think the logic can be simplified a bit along the lines of
```
def remove_foreign_key(oott)
  if oott.is_a?(Hash)
    name_or_column = oott.has_key?(:name) ? :name : :column
    foreign_keys.delete_if do |_, options|
      options[name_or_column].to_s == oott[name_or_column].to_s
    end
  elsif oott.is_a?(String) || oott.is_a?(Symbol)
    foreign_keys.delete(oott.to_s)
  else
    raise ArgumentError, "remove_foreign_key needs either a table name or a foreign key options hash with :name or :column"
  end    
end
```

(At least I think this version is a bit easier to read)

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 4]
```diff
@@ -309,6 +315,26 @@ def exec_rollback_db_transaction #:nodoc:
         log('rollback transaction',nil) { @connection.rollback }
       end
 
+      # REFERNTIAL INTEGRITY =====================================
```
Comment (timbreitkreutz) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Typo: `REFERENTIAL INTEGRITY`

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 5]
```diff
@@ -553,56 +614,177 @@ def sqlite_version
...
+
+        def table_structure(table_name)
+          structure = exec_query("PRAGMA table_info(#{quote_table_name(table_name)})", 'SCHEMA')
+          raise(ActiveRecord::StatementInvalid, "Could not find table '#{table_name}'") if structure.empty?
+
+          sql = "SELECT sql FROM sqlite_master WHERE type='table' and name=#{ quote(table_name) }"
           result = exec_query(sql, 'SCHEMA').first
+          unless result
+            return structure
+          end
```
Comment (timbreitkreutz) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Maybe
```
return structure unless result
```
?

**Comment by DavidMikeSimon** [comment]
@matthewd I strongly agree. :-( However I don't see any other way to get some meta-data (constraint names and collations) safely; I added a test to test/cases/adapters/sqlite3/collation_test.rb that demonstrates the bug in the current regex-based solution.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb (hunk 1 above)
Comment (DavidMikeSimon) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Okay. Actually, better yet, I'll have it ask SQLite with a PRAGMA whether or not foreign key support is working; it is possible for even recent versions of sqlite to be compiled with foreign key support deliberately disabled, though I don't know why anyone would do that.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb (hunk 2 above)
Comment (DavidMikeSimon) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: I removed it because `alter_table` is no longer using temporary tables. However, I should and will put it back in; I had forgotten that temporary tables can also be created directly by the user.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb (hunk 5 above)
Comment (nonmadden) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: :+1:

Comment (DavidMikeSimon) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Okay.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb (hunk 1 above)
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: A simple version check is fine. This doesn't need to execute a query.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb (hunk 3 above)
Comment (DavidMikeSimon) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb]: Yeah, your approach looks better to me as well, thank you. :-)

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb (hunk 4 above)
Comment (DavidMikeSimon) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Whoops, got it.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb (hunk 1 above)
Comment (DavidMikeSimon) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: I'll have it do the check and cache the result when it enables foreign keys on connection start.

**Comment by DavidMikeSimon** [comment]
Actually, maybe instead of getting rid of the regex I can harden it. The hard part there will be not creating unreadable regex-soup.

**Comment by DavidMikeSimon** [comment]
Okay, I think I've addressed all the suggestions so far. I replaced the parser with some fairly straightforward (albeit verbose) regex; the only particularly weird thing is that I use the `\g<expr>` syntax to handle nested parentheses correctly.

**Comment by DavidMikeSimon** [comment]
Rebased on latest master.

**Comment by DavidMikeSimon** [comment]
@smathy @timbreitkreutz @matthewd @sgrif Any further changes I should make?

**Comment by DavidMikeSimon** [comment]
Bump. Anything else I should do on this?

**Comment by smathy** [comment]
Looks good for me :+1:

**Comment by DavidMikeSimon** [comment]
Thanks @smathy. :-)

I've rebased on latest master.

**Comment by DavidMikeSimon** [comment]
Should I rebase again?

**Comment by DavidMikeSimon** [comment]
I'm going to close this for now, it's not feasible to rebase at this point and probably the *actual* correct solution starts with adding methods to the SQLite3 C API that make it easier to reflect on foreign keys.

**Comment by smathy** [comment]
Sorry to see all that very decent work sit here and go stale.  It can be very hard to get traction for non critical changes to essentially non-production adapters like sqlite :(

**Comment by DavidMikeSimon** [comment]
@smathy Thank you for your support. :-) OTOH I can also understand why there was hesitation about merging this PR: it was a large number of line changes for relatively little payoff.
